packages/integration-sdk-runtime/src/execution/duplicateKeyTracker.ts

```diff
@@ -0,0 +1,155 @@
+      );
+    }
+
+    this.graphObjectKeyMap.set(normalizedKey, metadata);
+  }
+
+  getGraphObjectMetadata(_key: string) {
+    return this.graphObjectKeyMap.get(this.normalizationFunction(_key));
+  }
+
+  hasKey(_key: string) {
+    return this.graphObjectKeyMap.has(this.normalizationFunction(_key));
+  }
+}
+
+type DuplicateKeyReportParams = {
+  duplicateEntity: Entity;
+  payload: Entity[];
+  indexOfDuplicateKey: number;
+  graphObjectStore: GraphObjectStore;
+};
+
+export async function createDuplicateEntityReport({
+  duplicateEntity,
+  payload,
+  indexOfDuplicateKey,
+  graphObjectStore,
+}: DuplicateKeyReportParams): Promise<DuplicateEntityReport> {
+  const originalEntity = originalInsidePayload(
+    payload,
+    duplicateEntity._key,
+    indexOfDuplicateKey,
+  );
+
+  if (originalEntity) {
+    return compareEntities(originalEntity, duplicateEntity);
+  } else {
+    const originalEntity = await graphObjectStore.findEntity(
+      duplicateEntity._key,
+    );
+    return compareEntities(originalEntity!, duplicateEntity);
+  }
+}
+
+/**
+ * Determines if the original entity that is duplicated by the new entity with the same key
+ * is inside the payload or in the graphObjectStore.
+ *
+ * @param payload the payload a duplicate _key was found in
+ * @param _key the _key that is a duplicate
+ * @param duplicateFoundIndex the index of the Entity or Relationship that triggered the DUPLICATE_KEY_ERROR
+ * @returns the original entity or relationship if it is inside the payload, otherwise returns undefined
+ */
+function originalInsidePayload(
+  payload: Entity[],
+  _key: string,
+  duplicateFoundIndex: number,
+): Entity | undefined {
+  const foundIndex = payload.findIndex((v: Entity) => {
+    return v._key === _key;
+  });
+
+  if (foundIndex === duplicateFoundIndex) {
+    return undefined;
+  } else {
+    return payload[foundIndex];
+  }
+}
+
+/**
+ * isDeepStrictEqual deeply compares two values and returns true if they are equal
+ * @param a any
+ * @param b any
+ * @returns boolean representing if the two values are deeply equal
+ */
+function isDeepStrictEqual(a: any, b: any): boolean {
+  try {
+    deepStrictEqual(a, b);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+type DuplicateEntityReport = {
+  _key: string;
+  rawDataMatch: boolean;
+  propertiesMatch: boolean;
+  fullMatch: boolean;
+};
+
+/**
+ * compareEntities compares two entities and produces a DuplicateEntityReport describing their
+ * similarities and differences.
+ * @param a
+ * @param b
+ * @returns
+ */
+function compareEntities(a: Entity, b: Entity): DuplicateEntityReport {
+  const aClone = JSON.parse(JSON.stringify(a));
+  const bClone = JSON.parse(JSON.stringify(b));
+  aClone._rawData = undefined;
+  bClone._rawData = undefined;
+
+  return {
+    _key: a._key,
+    rawDataMatch: isDeepStrictEqual(a._rawData, b._rawData),
+    propertiesMatch: isDeepStrictEqual(aClone, bClone),
+    fullMatch: isDeepStrictEqual(a, b),
```

Comment:
Is this inherently true if the above two properties are true? If so, we should remove this.